### PR TITLE
tests: smp: Add tests to verify SMP

### PR DESCRIPTION
**Pull request by spoorthik**
Added tests to verify scheduling behavior with SMP enabled
and added Ztest framework.

Signed-off-by: Spoorthi K <spoorthi.k@intel.com>

### CODE REVIEW

File: tests/kernel/smp/src/main.c [hunk 1]
```diff
@@ -83,9 +84,48 @@ void test_main(void)
...
+ * @details Verify whether thread running on other core is
+ * parent thread from child thread
+ */
+void test_cpu_id_threads(void)
+{
+	int parent_cpu_id = _arch_curr_cpu()->id;
+
+	k_thread_create(&t2, t2_stack, T2_STACK_SIZE, child_fn,
+			(void *)parent_cpu_id, &t2,
+			NULL, K_PRIO_PREEMPT(2), 0, K_NO_WAIT);
```
Comment (andyross) [inline tests/kernel/smp/src/main.c]: This looks racy.  What happens if test_cpu_id_threads finishes and exits (which it does synchronously as soon as k_thread_create() returns) before child_fn starts running on the other CPU?  Seems not at all unlikely, and will produce a spurious failure.

You need synchronization here.  Block on a semaphore that is given in child_fn, for example.

File: tests/kernel/smp/src/main.c [hunk 2]
```diff
@@ -83,9 +84,48 @@ void test_main(void)
...
+static void child_fn(void *p1, void *p2, void *p3)
+{
+	ARG_UNUSED(p3);
+	int parent_cpu_id = (int)p1;
+	struct k_thread *parent_thread_id = (struct k_thread *)p2;
+
+	zassert_true(parent_cpu_id != _arch_curr_cpu()->id &&
+		     _kernel.cpus[parent_cpu_id].current
+		     == parent_thread_id,
+		     "Parent isn't on other core");
```
Comment (andyross) [inline tests/kernel/smp/src/main.c]: This is going to be as reliable as you hope.  You don't know for sure that there aren't other threads (work queues, maybe) that might be preempting either CPU.  There probably aren't right now, in this test, but tests get augmented and moved around all the time and in some of the bigger ones there are MANY "forgotten" threads floating around from "finished" tests that still have a little work to do (and like to exercise scheduler bugs, heh!).

Broadly: Whiteboxing _kernel internals isn't exactly unacceptable in tests, but I don't see why this is needed here.  If you want to see that the other CPU is running the thread you expect, have it do something.  For example you could have one spin on an atomic flag variable (just a volatile global int is fine) waiting for a "message" from the other CPU, then set some output flag in a particular way.  Then you can know what is happening on the other CPU without mucking with internals.

**Review by andyross — CHANGES_REQUESTED** [review]
Scope is much smaller.  I actually liked all those other test cases, where'd they go?

One synchronization glitch that I can see needs fixing, and one note about implementation strategies you might want to consider.  The rest looks fine.

**Comment by spoorthik** [comment]
@andyross: Thanks for the review comments. I will work towards them and get back.

**Comment by spoorthik** [comment]
@andyross : Would you please have a look at new patch?

**Review by andyross — APPROVED** [review]
Looks good.  Thanks.

**Pull request merged**